lib/galaxy/model/__init__.py

```diff
@@ -10233,6 +10237,111 @@ class CleanupEventImplicitlyConvertedDatasetAssociationAssociation(Base):
     icda_id = Column(Integer, ForeignKey("implicitly_converted_dataset_association.id"), index=True)
 
 
+class CeleryUserRateLimit(Base):
+    """
+    For each user stores the last time a task was scheduled for execution.
+    Used to limit the number of tasks allowed per user per second.
+    """
+
+    __tablename__ = "celery_user_rate_limit"
+
+    user_id = Column(Integer, ForeignKey("galaxy_user.id", ondelete="CASCADE"), primary_key=True)
+    last_scheduled_time = Column(DateTime, nullable=False)
+
+    def calculate_task_start_time(
+        self, user_id: int, sa_session: Session, task_interval_secs: float, now: datetime.datetime
+    ) -> datetime.datetime:
+        """
+        Calculates the next time a task should be scheduled to run for the user_id.
+        It looks up the latest time a task has already been scheduled to run
+        for this user in a database table and then adds task_interval_secs seconds
+        to it to come up with the next scheduled time.
+        """
+        return now
+
+    def __repr__(self):
+        return (
+            f"CeleryUserRateLimit(id_type={self.id_type!r}, "
+            f"id={self.id!r}, last_scheduled_time={self.last_scheduled_time!r})"
+        )
+
+
+class CeleryUserRateLimitStandard(CeleryUserRateLimit):
+    """
+    Generic but slower implementation supported by most databases
+    """
+
+    _select_stmt = (
+        select(CeleryUserRateLimit.last_scheduled_time)
+        .with_for_update(of=CeleryUserRateLimit.last_scheduled_time)
+        .where(CeleryUserRateLimit.user_id == bindparam("userid"))
+    )
+
+    _update_stmt = (
+        update(CeleryUserRateLimit)
+        .where(CeleryUserRateLimit.user_id == bindparam("userid"))
+        .values(last_scheduled_time=bindparam("sched_time"))
+    )
+
+    _insert_stmt = insert(CeleryUserRateLimit).values(
+        user_id=bindparam("userid"), last_scheduled_time=bindparam("sched_time")
+    )
+
+    def calculate_task_start_time(
```

Comment:
This applies to both of the `calculate_task_start_time` methods.

Since the executed statement writes to the database, we need to explicitly commit, otherwise a rollback will be issued on session close (the session is no longer autocommited after #15421).

You can use this code: (the use of the context manager is a precaution during the transition to no autocommit and SQLAlchemy 2.0)

```py
from galaxy.model.base import transaction
...
with transaction(sa_session):
    sa_session.commit()
```